pisa/stages/xsec/pi_genie_sys.py

```diff
@@ -0,0 +1,230 @@
+    """
+    Stage to apply pre-calculated Genie systematics.
+
+    Paramaters
+    ----------
+    data
+    params
+        Must contain ::
+
+            Genie_Ma_QE : quantity (dimensionless)
+            Genie_Ma_RES : quantity (dimensionless)
+
+    input_names
+    output_names
+    debug_mode
+    input_specs
+    calc_specs
+    output_specs
+
+    Notes
+    -----
+    Requires the events have the following keys :
+
+        GENIE_linear_fit_MaCCQE
+            Genie CC quasi elastic linear coefficient
+        GENIE_quad_fit_MaCCQE
+            Genie CC quasi elastic quadratic coefficient
+        GENIE_linear_fit_MaCCRES
+            Genie CC resonance linear coefficient
+        GENIE_quad_fit_MaCCRES
+            Genie CC resonance quadratic coefficient
+
+    """
+    def __init__(
+        self,
+        data=None,
+        params=None,
+        input_names=None,
+        output_names=None,
+        debug_mode=None,
+        input_specs=None,
+        calc_specs=None,
+        output_specs=None,
+    ):
+        expected_params = (
+            'Genie_Ma_QE',
+            'Genie_Ma_RES',
+            'Genie_AhtBY',
+            'Genie_BhtBY',
+            'Genie_CV1uBY',
+            'Genie_CV2uBY',
+            'nu_diff_DIS',
+            'nubar_diff_DIS',
+            'hadron_DIS',
+            'A_scale_DIS',
+        )
+        input_names = ()
+        output_names = ()
+
+        # what are the keys used from the inputs during apply
+        input_apply_keys = (
+            'GENIE_linear_fit_MaCCQE',
+            'GENIE_quad_fit_MaCCQE',
+            'GENIE_linear_fit_MaCCRES',
+            'GENIE_quad_fit_MaCCRES',
+            'GENIE_linear_fit_AhtBY',
+            'GENIE_quad_fit_AhtBY',
+            'GENIE_linear_fit_BhtBY',
+            'GENIE_quad_fit_BhtBY',
+            'GENIE_linear_fit_CV1uBY',
+            'GENIE_quad_fit_CV1uBY',
+            'GENIE_linear_fit_CV2uBY',
+            'GENIE_quad_fit_CV2uBY',            
+            'GENIE_x',            
+            'GENIE_y',            
+            'GENIE_A',            
+        )
+        # what keys are added or altered for the outputs during apply
+        output_apply_keys = (
+            'weights',
+        )
+
+        # init base class
+        super(pi_genie_sys, self).__init__(
+            data=data,
+            params=params,
+            expected_params=expected_params,
+            input_names=input_names,
+            output_names=output_names,
+            debug_mode=debug_mode,
+            input_specs=input_specs,
+            calc_specs=calc_specs,
+            output_specs=output_specs,
+            input_apply_keys=input_apply_keys,
+            output_apply_keys=output_apply_keys,
+        )
+
+        assert self.input_mode is not None
+        assert self.calc_mode is None
+        assert self.output_mode is not None
+
+    @profile
+    def apply_function(self):
+        genie_ma_qe = self.params.Genie_Ma_QE.m_as('dimensionless')
```

Comment:
Avoid so much copy-pasting by defining somewhere the arguments to your function and then you can use a dictionary comprehension. E.g., since you've already defined the things you need elsewhere, you don't have to have so much effectively redundant code:

```python
kwargs = {name.lower(): getattr(self.params, name).m_as("dimensionless") for name in self.expected_params}
for container in self.data:
    for name in self.input_apply_keys:
        kwargs[name] = container[name].get(WHERE)
    apply_genie_sys(out=container['weights'].get(WHERE), **kwargs)
```